### PR TITLE
Docker deployment

### PR DESCRIPTION
**Pull request by laurenfrederick**

### CODE REVIEW

**Review by flamingbear — CHANGES_REQUESTED** [review]
This is freaking awesome.  I had a problem with teacache that I couldn't fix, so I'm going to try again tonight. But wanted to get your input on a couple of things and one correction.  But 😍

File: docker/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+## How to deploy
+
+### Set environment variables
+
+The deployment uses locally defined environment variables to generate all of the variables used for deployments. Define the following variables:
+
+- AWS_ACCOUNT_ID
+- AWS_ACCESS_KEY_ID
+- AWS_SECRET_ACCESS_KEY
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:27]: Yep, good point. Thanks for the suggestion. I have committed it.

File: docker/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+- RDS Cluster
+- Data persistence - database tables and Elasticsearch
+- Data migration
+- Cumulus
+
+### Connect to backend API
+
+`sh print-connection-commands.sh`
+
+This will print out something l
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:83]: Yeah not sure how that got deleted. Sorry.

File: docker/README.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+The build command:
+- Installs requirements (`requirements.sh`)
+- Copies deployment files to the docker container in the /deploy folder
+- Sets up your AWS credentials
+- Creates a Terraform state bucket if it does not exist
+- Initializes the RDS cluster deployment, the data persistence deployment, the data migration deployment, and the cumulus deployment
+  - Generates the `terraform.tf` file based on environment variables
+  - Generates the `terraform.tfvars` file based on environment variables
+  - Runs `terraform init`
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:47]: I added a few things. See what you think.

File: docker/README.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+This bare bones deployment does not include:
+- CMR integration
+- NASA Launchpad
+- Data direct S3 distribution access
+- Locking your Terraform state as you deploy
+
+## Prerequisites
+
+- Docker
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:13]: Fixed.

File: docker/README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+- Creates a Terraform state bucket if it does not exist
+- Initializes the RDS cluster deployment, the data persistence deployment, the data migration deployment, and the cumulus deployment
+  - Generates the `terraform.tf` file based on environment variables
+  - Generates the `terraform.tfvars` file based on environment variables
+  - Runs `terraform init`
+
+### Initialize a bash shell in your container
+
+`docker-compose run deploy /bin/bash`
+
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:52]: Added some wording to where you start the shell script inside of the container.

File: docker/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/README.md:83]: Fixed.

File: docker/prep-deployments.sh [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+sed -e 's/var.vpc_id/data.aws_vpc.application_vpcs.id/g' main.tf >> main.tf.tmp && mv main.tf.tmp main.tf
+sed -e 's/var.subnets/data.aws_subnet_ids.subnet_ids.ids/g' main.tf >> main.tf.tmp && mv main.tf.tmp main.tf
+
+../../terraform init
+
+echo "
+  cluster_identifier = \"$PREFIX-cumulus-rds-serverless\"
+  db_admin_username = \"$PREFIX"username"\"
+  db_admin_password = \"$PREFIX-password\"
+  rds_user_password = \"$PREFIX"password"\"
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/prep-deployments.sh]: No, they can't be or you get an error. I actually had to push up a fix for when there is a hyphen in the prefix. eek.

File: docker/prep-deployments.sh [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    bucket = \"$TFSTATE_BUCKET\"
+    key    = \"$CUMULUS_KEY\"
+    region = \"$AWS_REGION\"
+  }
+}" >> terraform.tf
+
+# Replace values in main.tf with the values coming from vpcConfig.tf
+sed -e 's/var.vpc_id/data.aws_vpc.application_vpcs.id/g' main.tf >> main.tf.tmp && mv main.tf.tmp main.tf
+sed -e 's/var.lambda_subnet_ids/data.aws_subnet_ids.subnet_ids.ids/g' main.tf >> main.tf.tmp && mv main.tf.tmp main.tf
+# sed -e 's/var.ecs_cluster_instance_image_id/data.aws_ssm_parameter.ecs_image_id.value/g' main.tf >> main.tf.tmp && mv main.tf.tmp main.tf
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/prep-deployments.sh:188]: yep

File: docker/prep-deployments.sh [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    bucket = \"$TFSTATE_BUCKET\"
+    key    = \"$DATA_PERSISTENCE_KEY\"
+    region = \"$AWS_REGION\"
+  }
+  provider_kms_key_id=\"$provider_kms_key\"
+  rds_user_access_secret_arn = \"\"
+  vpc_id=\"\"
+" >> terraform.tfvars
+
+# ../../terraform apply -auto-approve -input=false >> ../../output/data-migration-deployment.txt
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/prep-deployments.sh:161]: 👍

File: docker/teardown.sh [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+cd ../rds-cluster-tf
+
+../../terraform destroy -auto-approve -input=false
+
+### TEARDOWN BUCKETS ###
+
+aws s3 rb s3://$PREFIX-tf-state --force
+aws s3 rb s3://$PREFIX-internal --force
+aws s3 rb s3://$PREFIX-public
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/teardown.sh]: yes I fixed it a little bit ago

File: docker/vpcConfig.tf [hunk 10]
```diff
@@ -0,0 +1,14 @@
+data "aws_vpc" "application_vpcs" {
```
[the remark this answers is not in the document]
Comment (laurenfrederick) [reply docker/vpcConfig.tf:1]: It's finding a VPC with the name "Application VPC", which we're relying on being an NGAP standard here and then creating a VPC resource when it finds that. For the subnet it's using a tag name and again we're relying on the NGAP standard for the tag.

**Review by flamingbear — APPROVED** [review]

**Comment by flamingbear** [comment]
This looks great now.

**Review by nemreid — CHANGES_REQUESTED** [review]
Blocking on potential concern about resources left behind from teardown.

File: docker/README.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+`sh print-connection-commands.sh`
+
+This will print out something like:
+```
+# API Connection via tunneling
+## Create local port:
+aws ssm start-session --target i-000000000000 --document-name AWS-StartPortForwardingSession --parameters portNumber=22,localPortNumber=4343
+
+## Run in a separate terminal to create the tunnel:
```
Comment (nemreid) [inline docker/README.md:98]:
```suggestion
## Configure with path to keypair and run in a separate terminal to create the tunnel:
```

File: docker/README.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+ssh -p 4343 -N -L 8000:abcdefghij.execute-api.us-east-1.amazonaws.com:443 ec2-user@localhost
```
Comment (nemreid) [inline docker/README.md:99]:
```suggestion
ssh -p 4343 -N -L 8000:abcdefghij.execute-api.us-east-1.amazonaws.com:443 ec2-user@localhost -i path/to/keypair
```

File: docker/print-connection-commands.sh [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+api_gateway=$(../../terraform output | grep "archive_api_uri" | sed 's/.*https:\/\/\(.*\):.*/\1/')
+
+api_gateway_redirect_uri=$(../../terraform output | grep "archive_api_redirect_uri" | sed 's/.* = \(.*\)/\1/')
+
+echo "
+# API Connection via tunneling
+## Create local port:
+aws ssm start-session --target $ec2_instance_id --document-name AWS-StartPortForwardingSession --parameters portNumber=22,localPortNumber=4343
+
+## Run in a separate terminal to create the tunnel:
```
Comment (nemreid) [inline docker/print-connection-commands.sh:19]:
```suggestion
## Configure with path to keypair and run in a separate terminal to create the tunnel:
```

File: docker/print-connection-commands.sh [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+ssh -p 4343 -N -L 8000:$api_gateway:443 ec2-user@localhost
```
Comment (nemreid) [inline docker/print-connection-commands.sh:20]:
```suggestion
ssh -p 4343 -N -L 8000:$api_gateway:443 ec2-user@localhost -i path/to/keypair
```

**Review by nemreid — APPROVED** [review]
👍

**Pull request merged**